File: 125.Valid_Palindrome/main.cpp
#include <iostream>
#include <cctype>
#include <string>

class ValidPalindrome
{
public:
     bool isPalindrome(const std::string &s)
     {
          int left = 0, right = s.size() - 1;
          while (left < right)
          {
               while (left < right && !isAlphanumeric(s[left]))
                    left++;
               while (left < right && !isAlphanumeric(s[right]))
                    right--;
               if (toLower(s[left]) != toLower(s[right]))
               {
                    return false;
               }
               left++;
               right--;
          }
          return true;
     }

private:
     bool isAlphanumeric(char c)
     {
          return std::isalnum(c);
     }

     char toLower(char c)
     {
          return std::tolower(c);
     }
};

int main()
{
     ValidPalindrome validator;
     std::string s = "A man, a plan, a canal: Panama";
     std::cout << (validator.isPalindrome(s) ? "true" : "false") << std::endl; // Output: true
     return 0;
}

File: 125.Valid_Palindrome/main.py
class ValidPalindrome:
    def is_palindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            while left < right and not self.is_alphanumeric(s[left]):
                left += 1
            while left < right and not self.is_alphanumeric(s[right]):
                right -= 1
            if self.to_lower(s[left]) != self.to_lower(s[right]):
                return False
            left += 1
            right -= 1
        return True

    def is_alphanumeric(self, c: str) -> bool:
        return c.isalnum()

    def to_lower(self, c: str) -> str:
        return c.lower()

if __name__ == "__main__":
    validator = ValidPalindrome()
    s = "A man, a plan, a canal: Panama"
    print(validator.is_palindrome(s))  # Output: true
